crates/natrix_macros/src/lib.rs

```diff
@@ -24,6 +24,369 @@ pub fn format_elements(input: proc_macro::TokenStream) -> proc_macro::TokenStrea
     formatting::format_elements(input)
 }
 
+/// Derive the `Project` trait for an enum
+#[proc_macro_derive(Project, attributes(project))]
+pub fn project_derive(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let item = syn::parse_macro_input!(item as syn::ItemEnum);
+    let name = &item.ident;
+    let generics = &item.generics;
+    let (impl_generics, type_generics, where_clause) = generics.split_for_impl();
+
+    // Parse additional derives from attributes
+    let mut additional_derives = quote!();
+    for attr in &item.attrs {
+        if attr.path().is_ident("project") {
+            if let syn::Meta::List(meta_list) = &attr.meta {
+                // Extract the tokens and create a derive attribute
+                let tokens = &meta_list.tokens;
+                additional_derives = quote!(#[derive(#tokens)]);
+            }
+        }
+    }
+
+    // Generate projected enum generics with lifetime only if needed
+    let needs_lifetime = item.variants.iter().any(|variant| {
+        !matches!(variant.fields, syn::Fields::Unit)
+    });
```

Comment:
Honestly we could prolly... arghh

okay so I was gonna say we could honestly make a pure-unit enum emit compile_error.
actually we should, because in fact a pure-unit enum breaks some invariants (like theres no good way to preserve the `None` case of a failable ref)